### PR TITLE
Fix XML Reporters not Escaping Characters

### LINKED ISSUE #968: Checkstyle Reporter: message is not allowed to contain `<`
Example output:

```
<error line="28" column="18" severity="warning" message="Shorthand syntactic sugar should be used, i.e. [Int] instead of Array<Int>"/>
```

This seems not to be valid xml, QuickBuild or xml validators complain like this about it:

```
java.lang.RuntimeException: org.dom4j.DocumentException: Error on line 4 of document : The value of attribute "message" associated with an element type "error" must not contain the '<' character. Nested exception: The value of attribute "message" associated with an element type "error" must not contain the '<' character.
```

### PR DESCRIPTION
**Pull request by fabb**
Fixes #968.

### CODE REVIEW

File: Source/SwiftLintFramework/Reporters/String+XML.swift [hunk 1]
```diff
@@ -0,0 +1,25 @@
+//
+//  CheckstyleReporter.swift
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Reporters/String+XML.swift]: Also, there's an "Extensions" group/folder, this file probably belong there

Comment (fabb) [inline Source/SwiftLintFramework/Reporters/String+XML.swift]: good catch :-)

**Comment by marcelofabri** [comment]
Thanks for this! Could you add a changelog entry? Check https://github.com/realm/SwiftLint/blob/master/CONTRIBUTING.md#tracking-changes

File: Source/SwiftLintFramework/Reporters/CheckstyleReporter.swift [hunk 2]
```diff
@@ -25,11 +25,11 @@ public struct CheckstyleReporter: Reporter {
     }
 
     private static func generateForSingleViolation(_ violation: StyleViolation) -> String {
-        let file: String = violation.location.file ?? "<nopath>"
+        let file: String = violation.location.file?.escapedForXml() ?? "<nopath>"
```
Comment (jpsim) [inline Source/SwiftLintFramework/Reporters/CheckstyleReporter.swift]: `"<nopath>"` needs to be escaped: `(violation.location.file ?? "<nopath>").escapedForXml()`

File: Source/SwiftLintFramework/Reporters/HTMLReporter.swift [hunk 3]
```diff
@@ -120,7 +120,7 @@ public struct HTMLReporter: Reporter {
     private static func generateSingleRow(for violation: StyleViolation, at index: Int) -> String {
         let severity: String = violation.severity.rawValue.capitalized
         let location = violation.location
-        let file: String = location.file ?? ""
+        let file: String = location.file?.escapedForXml() ?? ""
```
Comment (jpsim) [inline Source/SwiftLintFramework/Reporters/HTMLReporter.swift]: we should probably use `"<nopath>"` here too for consistency with `CheckstyleReporter.swift`: `(location.file ?? "<nopath>").escapedForXml()`

File: Source/SwiftLintFramework/Reporters/JUnitReporter.swift [hunk 4]
```diff
@@ -19,11 +19,12 @@ public struct JUnitReporter: Reporter {
     public static func generateReport(_ violations: [StyleViolation]) -> String {
         return "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<testsuites><testsuite>" +
             violations.map({ violation in
-                let fileName = violation.location.file ?? "<nopath>"
+                let fileName = violation.location.file?.escapedForXml() ?? "<nopath>"
```
Comment (jpsim) [inline Source/SwiftLintFramework/Reporters/JUnitReporter.swift]: `"<nopath>"` needs to be escaped: `(violation.location.file ?? "<nopath>").escapedForXml()`

**Comment by fabb** [comment]
Sorry for not reading CONRIBUTING.md :facepalm:

File: Source/SwiftLintFramework/Reporters/CheckstyleReporter.swift (hunk 2 above)
Comment (fabb) [inline Source/SwiftLintFramework/Reporters/CheckstyleReporter.swift]: thanks, good catch! i added an example to the unit tests

File: Source/SwiftLintFramework/Reporters/HTMLReporter.swift (hunk 3 above)
Comment (fabb) [inline Source/SwiftLintFramework/Reporters/HTMLReporter.swift]: done

File: CHANGELOG.md [hunk 5]
```diff
@@ -61,6 +61,10 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#934](https://github.com/realm/SwiftLint/issues/934)
 
+* Fix XML reporters not escaping characters
```
Comment (jpsim) [inline CHANGELOG.md]: still missing trailing period, followed by two spaces: https://github.com/realm/SwiftLint/blob/master/CONTRIBUTING.md#tracking-changes

File: Source/SwiftLintFramework/Extensions/String+XML.swift [hunk 6]
```diff
@@ -0,0 +1,25 @@
+//
+//  String+XML.swift
+//  SwiftLint
+//
+//  Created by Fabian Ehrentraud on 12/12/16.
+//  Copyright © 2016 Realm. All rights reserved.
+//
+
+extension String {
+    func escapedForXml() -> String {
```
Comment (jpsim) [inline Source/SwiftLintFramework/Extensions/String+XML.swift]: would be nice if `XML` was capitalized: `escapedForXML()`

File: Source/SwiftLintFramework/Reporters/HTMLReporter.swift [hunk 7]
```diff
@@ -25,16 +25,25 @@ public struct HTMLReporter: Reporter {
...
+        let dateString = formatter.string(from: Date())
+        return generateReport(
+                violations,
+                swiftlintVersion: swiftlintVersion,
+                dateString: dateString
+        )
+    }
+
+    // swiftlint:disable function_body_length
+    // swiftlint:disable line_length
```
Comment (fabb) [inline Source/SwiftLintFramework/Reporters/HTMLReporter.swift]: i tried these 2 variants, but it didn't work:

```
    // swiftlint:disable function_body_length line_length
```

```
    // swiftlint:disable:next function_body_length
    // swiftlint:disable:next line_length
```

bug?

File: CHANGELOG.md (hunk 5 above)
Comment (fabb) [inline CHANGELOG.md]: maybe we should lint that file in the unit tests ;-)

File: Source/SwiftLintFramework/Reporters/HTMLReporter.swift (hunk 7 above)
Comment (jpsim) [inline Source/SwiftLintFramework/Reporters/HTMLReporter.swift]: yeah, could you please file a new ticket for this? I don't think my quick implementation from last week is 100% correct.

Comment (fabb) [inline Source/SwiftLintFramework/Reporters/HTMLReporter.swift]: sure, done: #976

Comment (jpsim) [inline Source/SwiftLintFramework/Reporters/HTMLReporter.swift]: Fixed in #978, thanks!

**Comment by jpsim** [comment]
Thanks for pushing this forward @fabb! I've made some small changes and addressed the merge conflicts so I'll be merging this as part of #979.

**Comment by fabb** [comment]
Awesome, thanks a lot!

**Pull request closed without merging**